### ISSUE TITLE
macOS x86_64 build error: __cpuid is unavailable on 10.7–10.8

### ISSUE THREAD

**Issue by barracuda156** [PurpleI2P/i2pd#1798]
I am in a process of bringing `i2pd` into Macports, and while the biggest issue is fixing PPC build, turned out that Intel is broken on 10.8.5 due to missing `__cpuid`.
```
:info:build /opt/local/var/macports/build/_opt_MountainLionPorts_net_i2pd/i2pd/work/i2pd-2.43.0/build/../libi2pd/Tunnel.h:98:9: note: hidden overloaded virtual function 'i2p::tunnel::Tunnel::SendTunnelDataMsg' declared here
:info:build                         void SendTunnelDataMsg (std::shared_ptr<i2p::I2NPMessage> msg);
:info:build                              ^
:info:build /opt/local/var/macports/build/_opt_MountainLionPorts_net_i2pd/i2pd/work/i2pd-2.43.0/libi2pd/CPU.cpp:34:3: error: use of undeclared identifier '__cpuid'
:info:build                 __cpuid(0, info[0], info[1], info[2], info[3]);
:info:build                 ^
:info:build /opt/local/var/macports/build/_opt_MountainLionPorts_net_i2pd/i2pd/work/i2pd-2.43.0/libi2pd/CPU.cpp:36:4: error: use of undeclared identifier '__cpuid'
:info:build                         __cpuid(0x00000001, info[0], info[1], info[2], info[3]);
:info:build                         ^
:info:build 2 errors generated.
```
See for more info: https://trac.macports.org/ticket/46038

P. S. Of course locally in Macports we can easily solve this by blacklisting Xcode clang on these two systems. But it will remain broken outside of Macports.

**Comment by r4sas** [comment]
This is strange because lines 34 and 36 should only be used if x86_64 or i386 is used. Is it cross compilation?

**Comment by r4sas** [comment]
Does changing CPU.cpp#32 to `#if (!defined __POWERPC__ && (defined(__x86_64__) || defined(__i386__))` help?

**Comment by barracuda156** [comment]
> This is strange because lines 34 and 36 should only be used if x86_64 or i386 is used. Is it cross compilation?

No, I have nothing to cross-compile for Intel on. My Intel builds are done in VMs running MacOS Intel on an Intel Macs.

For PPC I can build either natively or in Rosetta (still not a cross-compile case per se).

**Comment by r4sas** [comment]
Hm, okay. So that error happens on Intel, but only in specific OSX versions?

**Comment by barracuda156** [comment]
> Hm, okay. So that error happens on Intel, but only in specific OSX versions?

The issue appears to be compiler-based: several old versions of Clang fail that way. Ref: https://trac.macports.org/ticket/46038#comment:31
Why it translates into OS-based problem is because:

1. Clang starts being used in 10.6, but Clang of 10.6 does not support C++11 and therefore is not picked by the build, so macOS < 10.7 are not affected.
2. Clang of 10.9+ do not have that issue.

In result, there are two systems affected, where system Clang supports C++11, but has this bug, 10.7 and 10.8. At least, this is how I understand the issue from a failure I observed myself and reading related cases.
